File: src/utils/date.ts
/** 当前年月 YYYY-MM */
export function currentYearMonth() {
  const d = new Date()
  return `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, '0')}`
}

/** 月份加减 */
export function shiftYearMonth(ym: string, delta: number) {
  const [y, m] = ym.split('-').map(Number)
  const d = new Date(y, m - 1 + delta, 1)
  return `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, '0')}`
}

/** 月起止 ISO 日期 */
export function monthRange(ym: string) {
  const [y, m] = ym.split('-').map(Number)
  const start = `${ym}-01`
  const last = new Date(y, m, 0).getDate()
  const end = `${ym}-${String(last).padStart(2, '0')}`
  return { start, end }
}

/** 今天 YYYY-MM-DD */
export function todayDate() {
  const d = new Date()
  return `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, '0')}-${String(d.getDate()).padStart(2, '0')}`
}

/** 日分组标题：03月 12日 周二 */
export function formatDayLabel(isoDate: string) {
  const d = new Date(`${isoDate}T00:00:00`)
  const week = ['日', '一', '二', '三', '四', '五', '六'][d.getDay()]
  const month = String(d.getMonth() + 1).padStart(2, '0')
  const day = String(d.getDate()).padStart(2, '0')
  return `${month}月${day}日 周${week}`
}

/** 截止日期短展示：09月13日 */
export function formatDateShort(isoDate: string) {
  const d = new Date(`${isoDate}T00:00:00`)
  if (Number.isNaN(d.getTime())) return isoDate
  const month = String(d.getMonth() + 1).padStart(2, '0')
  const day = String(d.getDate()).padStart(2, '0')
  return `${month}月${day}日`
}

/** 展示用年月 */
export function formatYearMonth(ym: string) {
  const [y, m] = ym.split('-')
  return `${y}年${Number(m)}月`
}

/** 短年月：2019/2 */
export function formatYearMonthShort(ym: string) {
  const [y, m] = ym.split('-')
  return `${Number(m)}/${y}`
}

/** 从 YYYY-MM-DD 取年月 */
export function yearMonthOf(date: string) {
  return date.slice(0, 7)
}

/** 构建月历格子（含上月/下月占位） */
export function buildMonthGrid(ym: string) {
  const [y, m] = ym.split('-').map(Number)
  const first = new Date(y, m - 1, 1)
  const daysInMonth = new Date(y, m, 0).getDate()
  const startWeekday = first.getDay()
  const cells: Array<{
    date: string
    day: number
    inMonth: boolean
    isToday: boolean
  }> = []
  const today = todayDate()

  // 上月补齐
  const prevDays = new Date(y, m - 1, 0).getDate()
  for (let i = startWeekday - 1; i >= 0; i--) {
    const day = prevDays - i
    const pm = m === 1 ? 12 : m - 1
    const py = m === 1 ? y - 1 : y
    const date = `${py}-${String(pm).padStart(2, '0')}-${String(day).padStart(2, '0')}`
    cells.push({ date, day, inMonth: false, isToday: date === today })
  }

  for (let day = 1; day <= daysInMonth; day++) {
    const date = `${ym}-${String(day).padStart(2, '0')}`
    cells.push({ date, day, inMonth: true, isToday: date === today })
  }

  // 下月补齐到完整周
  let next = 1
  while (cells.length % 7 !== 0) {
    const nm = m === 12 ? 1 : m + 1
    const ny = m === 12 ? y + 1 : y
    const date = `${ny}-${String(nm).padStart(2, '0')}-${String(next).padStart(2, '0')}`
    cells.push({ date, day: next, inMonth: false, isToday: date === today })
    next += 1
  }

  return cells
}

/** 日标题短格式：15/2 */
export function formatDayShort(isoDate: string) {
  const d = new Date(`${isoDate}T00:00:00`)
  return `${d.getDate()}/${d.getMonth() + 1}/${d.getFullYear()}`
}

File: lib/core/utils/date.dart
/// 日期工具：年月运算、月历格子、展示文案。
library;

String currentYearMonth() {
  final d = DateTime.now();
  return '${d.year}-${d.month.toString().padLeft(2, '0')}';
}

String shiftYearMonth(String ym, int delta) {
  final parts = ym.split('-').map(int.parse).toList();
  final d = DateTime(parts[0], parts[1] + delta, 1);
  return '${d.year}-${d.month.toString().padLeft(2, '0')}';
}

({String start, String end}) monthRange(String ym) {
  final parts = ym.split('-').map(int.parse).toList();
  final last = DateTime(parts[0], parts[1] + 1, 0).day;
  return (
    start: '$ym-01',
    end: '$ym-${last.toString().padLeft(2, '0')}',
  );
}

String todayDate() {
  final d = DateTime.now();
  return '${d.year}-${d.month.toString().padLeft(2, '0')}-${d.day.toString().padLeft(2, '0')}';
}

class MonthDayItem {
  const MonthDayItem({
    required this.date,
    required this.day,
    required this.weekLabel,
  });

  final String date;
  final int day;
  final String weekLabel;
}

/// 生成当月横向日期条数据。
List<MonthDayItem> listMonthDays(String ym) {
  // 1. 解析年月并取当月天数
  // 2. 按日生成 ISO 日期与周几文案
  final parts = ym.split('-').map(int.parse).toList();
  final last = DateTime(parts[0], parts[1] + 1, 0).day;
  const week = ['日', '一', '二', '三', '四', '五', '六'];
  final days = <MonthDayItem>[];
  for (var day = 1; day <= last; day++) {
    final date = '$ym-${day.toString().padLeft(2, '0')}';
    // Dart weekday: Mon=1..Sun=7；对齐 JS getDay: Sun=0
    final jsWd = DateTime(parts[0], parts[1], day).weekday == 7
        ? 0
        : DateTime(parts[0], parts[1], day).weekday;
    days.add(MonthDayItem(date: date, day: day, weekLabel: '周${week[jsWd]}'));
  }
  return days;
}

String defaultDayInMonth(String ym) {
  final today = todayDate();
  if (today.startsWith('$ym-')) return today;
  return '$ym-01';
}

String formatDayLabel(String isoDate) {
  final d = DateTime.parse('${isoDate}T00:00:00');
  const week = ['日', '一', '二', '三', '四', '五', '六'];
  final wd = d.weekday == 7 ? 0 : d.weekday;
  final month = d.month.toString().padLeft(2, '0');
  final day = d.day.toString().padLeft(2, '0');
  return '$month月$day日 周${week[wd]}';
}

String formatDateShort(String isoDate) {
  final d = DateTime.tryParse('${isoDate}T00:00:00');
  if (d == null) return isoDate;
  return '${d.month.toString().padLeft(2, '0')}月${d.day.toString().padLeft(2, '0')}日';
}

String formatYearMonth(String ym) {
  final parts = ym.split('-');
  return '${parts[0]}年${int.parse(parts[1])}月';
}

String formatYearMonthShort(String ym) {
  final parts = ym.split('-');
  return '${int.parse(parts[1])}/${parts[0]}';
}

String yearMonthOf(String date) => date.substring(0, 7);

({String year, String month}) splitYearMonth(String ym) {
  final parts = ym.split('-');
  return (year: parts[0], month: '${int.parse(parts[1])}');
}

class MonthGridCell {
  const MonthGridCell({
    required this.date,
    required this.day,
    required this.inMonth,
    required this.isToday,
  });

  final String date;
  final int day;
  final bool inMonth;
  final bool isToday;
}

/// 构建月历格子（含上月/下月占位）。
List<MonthGridCell> buildMonthGrid(String ym) {
  final parts = ym.split('-').map(int.parse).toList();
  final y = parts[0];
  final m = parts[1];
  final first = DateTime(y, m, 1);
  final daysInMonth = DateTime(y, m + 1, 0).day;
  final startWeekday = first.weekday == 7 ? 0 : first.weekday;
  final cells = <MonthGridCell>[];
  final today = todayDate();

  final prevDays = DateTime(y, m, 0).day;
  for (var i = startWeekday - 1; i >= 0; i--) {
    final day = prevDays - i;
    final pm = m == 1 ? 12 : m - 1;
    final py = m == 1 ? y - 1 : y;
    final date =
        '$py-${pm.toString().padLeft(2, '0')}-${day.toString().padLeft(2, '0')}';
    cells.add(MonthGridCell(
      date: date,
      day: day,
      inMonth: false,
      isToday: date == today,
    ));
  }

  for (var day = 1; day <= daysInMonth; day++) {
    final date = '$ym-${day.toString().padLeft(2, '0')}';
    cells.add(MonthGridCell(
      date: date,
      day: day,
      inMonth: true,
      isToday: date == today,
    ));
  }

  var next = 1;
  while (cells.length % 7 != 0) {
    final nm = m == 12 ? 1 : m + 1;
    final ny = m == 12 ? y + 1 : y;
    final date =
        '$ny-${nm.toString().padLeft(2, '0')}-${next.toString().padLeft(2, '0')}';
    cells.add(MonthGridCell(
      date: date,
      day: next,
      inMonth: false,
      isToday: date == today,
    ));
    next += 1;
  }

  return cells;
}

String formatDayShort(String isoDate) {
  final d = DateTime.parse('${isoDate}T00:00:00');
  return '${d.day}/${d.month}/${d.year}';
}
